[AMDGPU] Remove the AnnotateKernelFeatures pass

jwanggit86 opened the pull request:
Previously the AnnotateKernelFeatures pass infers two attributes: amdgpu-calls and amdgpu-stack-objects, which are used to help determine if flat scratch init is allowed. PR #118907 created the amdgpu-no-flat-scratch-init attribute. Continuing with that work, this patch makes use of this attribute to determine flat scratch init, replacing amdgpu-calls and amdgpu-stack-objects. This also leads to the removal of the AnnotateKernelFeatures pass.

jwanggit86 committed:
[AMDGPU] Remove the AnnotateKernelFeatures pass

Previously the AnnotateKernelFeatures pass infers two attributes:
amdgpu-calls and amdgpu-stack-objects, which are used to help determine
if flat scratch init is allowed. PR #118907 created the
amdgpu-no-flat-scratch-init attribute. Continuing with that work, this
patch makes use of this attribute to determine flat scratch init,
replacing amdgpu-calls and amdgpu-stack-objects. This also leads to the
removal of the AnnotateKernelFeatures pass.

jwanggit86 committed:
Fix formatting.

shiltian commented on llvm/test/CodeGen/AMDGPU/GlobalISel/atomicrmw_udec_wrap.ll:
```diff
@@ -20,11 +20,14 @@ define amdgpu_kernel void @lds_atomic_dec_ret_i32(ptr addrspace(1) %out, ptr add
 ; CI-NEXT:    ds_dec_rtn_u32 v2, v1, v0
 ; CI-NEXT:    s_waitcnt lgkmcnt(0)
 ; CI-NEXT:    v_mov_b32_e32 v0, s0
+; CI-NEXT:    s_mov_b32 flat_scratch_lo, s13
```
Maybe add the attribute to those test cases to avoid updating these check lines

shiltian replied:
yes

arsenm reviewed (commented):
Should also add a test for what happens on undefined behavior, and flat_scratch does end up really getting used. We shouldn't have any verifier errors or anything

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/flat-scratch-init.ll:
```diff
@@ -35,15 +35,15 @@ define amdgpu_kernel void @stack_object_addrspacecast_in_kernel_no_calls() {
 ; RW-FLAT:     .amdhsa_user_sgpr_private_segment_buffer 1
 ; RO-FLAT-NOT: .amdhsa_user_sgpr_private_segment_buffer
-; RW-FLAT:     .amdhsa_user_sgpr_flat_scratch_init 1
+; RW-FLAT:     .amdhsa_user_sgpr_flat_scratch_init 0
```
This test looks doubly broken. For the intent of the test, this should be 1. It's also missing the attribute, so it definitely shouldn't be 0

jwanggit86 replied:
Previouly the AnnotateKernelFeatures pass checks for calls and allocas. With the change to AMDGPUAttributor, alloca is no longer being checked. For this particular case, opt would set the `no-flat-scratch-init` attribute. So after llc, flat_scratch_init in line 38 would be 0.

jwanggit86 commented:
> Should also add a test for what happens on undefined behavior, and flat_scratch does end up really getting used. We shouldn't have any verifier errors or anything

Please elaborate on "undefined behavior".

optimisan commented:
Could you also delete the entry for this pass from `AMDGPUPassRegistry.def`?

arsenm commented:
> > Should also add a test for what happens on undefined behavior, and flat_scratch does end up really getting used. We shouldn't have any verifier errors or anything
> 
> Please elaborate on "undefined behavior".

The contract of amdgpu-no-flat-scratch-init is violated and flat_scr requires initialization. i.e. addrspacecast of a private address to flat

jwanggit86 committed:
Update test files.

As amdgpu-no-flat-scratch-init is set by opt, in most llc tests the
attribute is not set. This commit corrects this.

jwanggit86 committed:
Remove the pass from AMDGPUPassRegistry.def

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/implicit-kernarg-backend-usage-global-isel.ll:
```diff
@@ -183,15 +177,12 @@ define amdgpu_kernel void @llvm_amdgcn_is_shared(ptr %ptr) {
 }
 
-define amdgpu_kernel void @llvm_amdgcn_is_private(ptr %ptr) {
+define amdgpu_kernel void @llvm_amdgcn_is_private(ptr %ptr) #0 {
```
This case is interesting because we could use the attribute to just fold this to a constant false. I think this should not get annotated

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/extractelement.ll:
```diff
@@ -4022,7 +4019,7 @@ entry:
 }
 
-define amdgpu_kernel void @dyn_extract_v4f32_s_s_s(ptr addrspace(1) %out, i32 %sel) {
+define amdgpu_kernel void @dyn_extract_v4f32_s_s_s(ptr addrspace(1) %out, i32 %sel) #0 {
```
The attribute isn't relevant enough to this test, I would just regenerate this with the new reference output

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/attributor-flatscratchinit-invalid.ll:
```diff
@@ -0,0 +1,550 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Test should be renamed this isn't testing attributor. How about amdgpu-no-flat-scratch-init-undefined-behavior.ll

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/attributor-flatscratchinit-invalid.ll:
```diff
@@ -0,0 +1,550 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s | FileCheck -check-prefixes=GFX9 %s
```
Add a gfx8 run line. It's more relevant than testing gfx10 since it has a different addrspacecast lowering. I would test gfx803, gfx900, and gfx942 for architected flat scratch

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/atomicrmw_udec_wrap.ll:
```diff
@@ -3319,7 +3319,7 @@ define amdgpu_kernel void @atomic_dec_shl_base_lds_0_i64(ptr addrspace(1) %out,
 
 attributes #0 = { nounwind speculatable willreturn memory(none) }
-attributes #1 = { nounwind }
+attributes #1 = { nounwind "amdgpu-no-flat-scratch-init"}
```
Not relevant to this test, just regenerate the results

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/insertelement-stack-lower.ll:
```diff
@@ -257,4 +257,4 @@ define amdgpu_kernel void @v_insert_v64i32_varidx(ptr addrspace(1) %out.ptr, ptr
 }
 
-attributes #0 = { "amdgpu-flat-work-group-size"="1,256" "amdgpu-waves-per-eu"="1,10" }
+attributes #0 = { "amdgpu-flat-work-group-size"="1,256" "amdgpu-waves-per-eu"="1,10" "amdgpu-no-flat-scratch-init" }
```
Not a relevant test

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/attributor-flatscratchinit-invalid.ll:
```diff
@@ -0,0 +1,550 @@
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1010 < %s | FileCheck -check-prefixes=GFX10 %s
+
+;
+; These functions all should have the attribute amdgpu-no-flat-scratch-init set if the AMDGPUAttributor
+; pass is run. Therefore the purpose is to test llc when the attribute is incorrectly missing.
```
This test is not running the attributor, only codegen. This is also the opposite of the test I wanted. I wanted to test the case where the functions do have the attribute, but violate the contract. That is, every function should already have the attribute set. And then use the casts and flat accesses which would require flat_scr to be initialized

jwanggit86 replied:
> This test is not running the attributor, only codegen. This is also the opposite of the test I wanted. I wanted to test the case where the functions do have the attribute, but violate the contract. That is, every function should already have the attribute set. And then use the casts and flat accesses which would require flat_scr to be initialized

Now I have two files, one to run opt (attributor-flatscratchinit-invalid.ll) and one to run llc (attributor-flatscratchinit-invalid2.ll). One question about opt. If a function is not supposed to have the attribute but somehow has it, should attributor remove it?

jwanggit86 replied:
> One question about opt. If a function is not supposed to have the attribute but somehow has it, should attributor remove it?
@arsenm What's your thought on this?

shiltian replied:
It should not have it at the first place, so the attributor should not remove it and we need to fix where it is added.

shiltian replied:
We used to run the attributor twice, which allowed some attributes to be added in between the two runs. Some of those are fine as they don't affect correctness. However, if any of the attributes added in that phase impact correctness, then that's a problem.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -629,12 +623,16 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
   if (ST.hasFlatAddressSpace() && AMDGPU::isEntryFunctionCC(CC) &&
       (IsAmdHsaOrMesa || ST.enableFlatScratch()) &&
-      (HasCalls || HasStackObjects || ST.enableFlatScratch()) &&
+      // The line below: If enableFlatScratch() is true, whether
+      // no-flat-scratch-init is set is not important. If enableFlatScratch()
+      // is false, FlatScratchInit cannot be true for graphics CC.
+      (ST.enableFlatScratch() ||
```
```suggestion
      // FlatScratchInit cannot be true for graphics CC.
      (ST.enableFlatScratch() ||
```

This sounds broken. Flat addressing should work fine in shaders. The only difference will be in the sequence used to initialize flat_scr needs to be different. But that's a problem for later.

Also can sink the hasFnAttribute down here.

arsenm replied:
I mean inline in the expression. Also "cannot be true", rephrase are something about the graphics shaders not having the input

jwanggit86 replied:
> I mean inline in the expression (and swap the order, CC check first). Also "cannot be true", rephrase are something about the graphics shaders not having the input

What does "not having the input" mean?

arsenm reviewed (commented):
Most of the test changes should just regenerate to use the new output, and not add the attribute. The attribute should only be added in cases where it's directly relevant to the test (like the kernel metadata tests)

jwanggit86 commented:
> Could you also delete the entry for this pass from `AMDGPUPassRegistry.def`?

Done.

shiltian commented:
> Most of the test changes should just regenerate to use the new output, and not add the attribute. The attribute should only be added in cases where it's directly relevant to the test (like the kernel metadata tests) 

I'm not sure. Adding the attribute to existing test cases can avoid unnecessary check line changes, which makes the line history more consistent.

jwanggit86 committed:
Update tests.

Undo the changes in previous commit. Now the amdgpu-no-flatscratch-init
attribute is only manually added if the tests are relevant to the
attribute.

jwanggit86 committed:
Update tests.

Renamed attributor-flatscratchinit-invalid.ll to
attributor-flatscratchinit-undefined-behavior.ll. Added RUN lines.

shiltian reviewed (approved):
LGTM but let's wait for @arsenm for the comment about graphics CC.

arsenm commented on llvm/test/CodeGen/AMDGPU/attributor-flatscratchinit-undefined-behavior.ll:
```diff
@@ -0,0 +1,63 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
I don't see the point of this test, we don't need to check what the attributor does in these cases. The content is also duplicated with the 2 version. Should drop this version?

jwanggit86 replied:
This one runs opt with the Attributor pass. The "2" version runs only llc.

jwanggit86 committed:
Minor change based on review.

jwanggit86 committed:
Formatting.

jwanggit86 commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -629,12 +623,16 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
+      // FlatScratchInit cannot be true for graphics CC if enableFlatScratch()
+      // is false.
+      (ST.enableFlatScratch() ||
+       (!AMDGPU::isGraphics(CC) &&
```
Checking on CC is moved to the front.

jwanggit86 commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -629,12 +623,16 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
       (IsAmdHsaOrMesa || ST.enableFlatScratch()) &&
-      (HasCalls || HasStackObjects || ST.enableFlatScratch()) &&
+      // FlatScratchInit cannot be true for graphics CC if enableFlatScratch()
+      // is false.
```
Comment is updated.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSubtarget.cpp:
```diff
@@ -629,12 +623,16 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
-  // TODO: This could be refined a lot. The attribute is a poor way of
-  // detecting calls or stack objects that may require it before argument
-  // lowering.
+  const bool IsNoFlatScratchInitSet =
+      F.hasFnAttribute("amdgpu-no-flat-scratch-init");
+
```
```suggestion
```

This is now unused

jwanggit86 replied:
Removed.

jwanggit86 committed:
(1) remove useless code (2) update one test file.

jwanggit86 committed:
Trigger Build
